language/diem-framework/modules/0L/Teams.move

```diff
@@ -0,0 +1,302 @@
+      collective_tower_height_this_epoch: u64,
+
+      // Informational fields
+      team_name: vector<u8>, // A validator account.
+      description: vector<u8>,
+      count_all_members: u64, // count the miners who have expressed to me members of this team, but may be below threshold
+      count_active: u64, // the count of players above threshold
+
+      // Is team in validator set
+    }
+
+    // this struct is stored in the member's account
+    struct Member has key, copy, drop, store {
+      captain_address: address, // by address of captain
+      mining_above_threshold: bool, // if the mining the user has done is above the system threshold to count toward delegation.
+
+    }
+
+    public fun vm_init(sender: &signer) {
+      CoreAddresses::assert_vm(sender);
+      if (!exists<AllTeams>(CoreAddresses::VM_RESERVED_ADDRESS())) {
+        move_to<AllTeams>(
+          sender, 
+          AllTeams {
+            teams_list: Vector::empty(),
+            collective_threshold_epoch: 0,
+            member_threshold_epoch: 0,
+            tower_height_rms: 0,
+          }
+        );
+      }
+    }
+
+
+    public fun team_init(sender: &signer, team_name: vector<u8>, operator_pct_reward: u64) {
+
+      assert(ValidatorUniverse::is_in_universe(Signer::address_of(sender)), 201301001);
+      // An "captain", who is already a validator account, stores the Team struct on their account.
+      // the AllTeams struct is saved in the 0x0 account, and needs to be initialized before this is called.
+
+      // check vm has initialized the struct, otherwise exit early.
+      if (!exists<AllTeams>(CoreAddresses::VM_RESERVED_ADDRESS())) {
+        return
+    };
+
+    move_to<Team>(
+        sender, 
+        Team {
+          captain: Signer::address_of(sender), // A validator account.
+          members: Vector::empty<address>(),
+          operator_pct_reward, // the percentage of the rewards that the captain proposes to go to the validator operator.
+          collective_tower_height_this_epoch: 0,
+
+          team_name, // A validator account.
+          description: Vector::empty<u8>(), // TODO: Change this
+          count_all_members: 0,
+          count_active: 0,
+
+        }
+      );
+    }
+
+
+    // a member states his team's preference.
+    // the member only actually joins a team for the purposes of calculating rewards
+    // when they do enough mining in a period
+    public fun join_team(sender: &signer, captain_address: address) acquires Member {
+      let addr = Signer::address_of(sender);
+      // needs to check if this is a slow wallet.
+      // ask user to resubmit if not a slow wallet, so they are explicitly setting it, no surprises, no tears.
+
+      assert(DiemAccount::is_slow(addr), ENOT_SLOW_WALLET);
+        
+
+      // bob wants to switch to a different Team.
+      if (exists<Member>(addr)) {
+        let member_state = borrow_global_mut<Member>(addr);
```

Comment:
Need to check if the account is in the member list of the old team and if so, remove them

Reply:
I think switch_team would allow a user to be added a team without having first met the mining threshold that is enforced is maybe_activate_member _to_team. Perhaps maybe_switch_team should call maybe_activate_member_to_team in order to prevent this?

Reply:
@jamesmeijers please check, I refactored as you suggested.  See: 12aa63f9a1377699f9f77499304271a05d652caf